06_puppy/CameronSchafer/main.go

```diff
@@ -0,0 +1,122 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"sync" //https://golang.org/pkg/sync/#Map
+)
+
+var out io.Writer = os.Stdout
+
+type Puppy struct {
+	pid    int
+	breed  string
+	colour string
+	value  string
+}
+
+type MapStore struct {
+	store map[int]Puppy
+}
+
+type SyncStore struct {
+	store sync.Map
+}
+
+type Storer interface {
+	CreatePuppy(*Puppy) int
+	ReadPuppy(int) *Puppy
+	UpdatePuppy(*Puppy)
+	DeletePuppy(int)
+}
+
+func (m *MapStore) CreatePuppy(p *Puppy) int {
+	m.store[p.pid] = *p
```

Comment:
IMO Create Puppy should provide the ID (the client/user of your package typically would not know what ID's are available). However you can make a case for the client providing the ID if you wish. In that case you need to check that the ID isn't already in use and return an error otherwise please.